unit_tests/test_provides.py

```diff
@@ -0,0 +1,130 @@
+import unittest
+import mock
+
+
+with mock.patch('charmhelpers.core.hookenv.metadata') as _meta:
+    _meta.return_Value = 'ss'
+    import provides
+
+
+_hook_args = {}
+
+TO_PATCH = [
+]
+
+
+def mock_hook(*args, **kwargs):
+
+    def inner(f):
+        # remember what we were passed.  Note that we can't actually determine
+        # the class we're attached to, as the decorator only gets the function.
+        _hook_args[f.__name__] = dict(args=args, kwargs=kwargs)
+        return f
+    return inner
+
+
+class _unit_mock:
+    def __init__(self, unit_name, received=None):
+        self.unit_name = unit_name
+        self.received = received or {}
+
+
+class _relation_mock:
+    def __init__(self, application_name=None, units=None):
+        self.to_publish_raw = {}
+        self.application_name = application_name
+        self.units = units
+
+
+class TestCinderBackendProvides(unittest.TestCase):
+
+    @classmethod
+    def setUpClass(cls):
+        cls._patched_hook = mock.patch('charms.reactive.when', mock_hook)
+        cls._patched_hook_started = cls._patched_hook.start()
+        # force provides to rerun the mock_hook decorator:
+        # try except is Python2/Python3 compatibility as Python3 has moved
+        # reload to importlib.
+        try:
+            reload(provides)
+        except NameError:
+            import importlib
+            importlib.reload(provides)
+
+    @classmethod
+    def tearDownClass(cls):
+        cls._patched_hook.stop()
+        cls._patched_hook_started = None
+        cls._patched_hook = None
+        # and fix any breakage we did to the module
+        try:
+            reload(provides)
+        except NameError:
+            import importlib
+            importlib.reload(provides)
+
+    def patch(self, method):
+        _m = mock.patch.object(self.obj, method)
+        _mock = _m.start()
+        self.addCleanup(_m.stop)
+        return _mock
+
+    def setUp(self):
+        self.cr = provides.CinderBackendProvides('some-relation', [])
+        self._patches = {}
+        self._patches_start = {}
+        self.obj = provides
+        for method in TO_PATCH:
+            setattr(self, method, self.patch(method))
+
+    def tearDown(self):
+        self.cr = None
+        for k, v in self._patches.items():
+            v.stop()
+            setattr(self, k, None)
+        self._patches = None
+        self._patches_start = None
+
+    def patch_kr(self, attr, return_value=None):
```

Comment:
nit: the origin of this is probably the keystone interface unit tests and I think the abbreviation `kr` refers to "keystone relation".  Would it be appropriate to either change the abbreviation to match constructs in this interface or pick a generic term we can re-use across interface unit tests?

Reply:
Good point. Should be fixed now.